solarforecastarbiter/reports/template.py

```diff
@@ -0,0 +1,121 @@
+"""
+Inserts metadata and figures into the report template.
+"""
+import subprocess
+
+from bokeh.embed import components
+
+from jinja2 import (Environment, DebugUndefined, PackageLoader,
+                    select_autoescape, Template)
+
+from solarforecastarbiter.reports import figures
+
+
+def prereport(metadata, metrics):
+    """
+    Render the markdown prereport. Figures are left untemplated.
+
+    Parameters
+    ----------
+    metadata : dict
+    metrics : dict
+
+    Returns
+    -------
+    prereport : markdown
+    """
+    # By default, jinja removes undefined variables from the rendered string.
+    # DebugUndefined leaves undefined variables in the string so that they
+    # can be used in the full report template process.
+    env = Environment(
+        loader=PackageLoader('solarforecastarbiter.reports', 'templates'),
+        autoescape=select_autoescape(['html', 'xml']),
+        undefined=DebugUndefined)
+
+    template = env.get_template('template.md')
+
+    print_versions = metadata['versions']
+    rendered = template.render(
+        name=metadata['name'],
+        start=metadata['start'],
+        end=metadata['end'],
+        now=metadata['now'],
+        print_versions=print_versions)
+    return rendered
+
+
+# not all args currently used, but expect they will eventually be used
+def add_figures_to_prereport(fx_obs_cds, report, metadata, prereport):
+    """
+    Add figures to the prereport, convert to html
+
+    Parameters
+    ----------
+    fx_obs_cds : list
+        List of (forecast, observation, ColumnDataSource) tuples to
+        pass to bokeh plotting objects.
+    report : solarforecastarbiter.datamodel.Report
+        Metadata describing report
+    metadata : str, json
+        Describes the pre-report
+    prereport : str, markdown or html
+        The templated pre-report.
+
+    Returns
+    -------
+    body : str, markdown or html
+        The body of the full report in the same format
+        (markdown or html) as the prereport.
+    """
+    body_template = Template(prereport)
+
+    ts_fig = figures.timeseries(fx_obs_cds, report.start, report.end)
+    scat_fig = figures.scatter(fx_obs_cds)
+
+    script, divs = components(
+        {'figures_timeseries': ts_fig, 'figures_scatter': scat_fig})
+
+    body = body_template.render(script=script, **divs)
+    return body
+
+
+def prereport_to_html(prereport):
+    """
+    Render markdown into simple html using pandoc.
+
+    Parameters
+    ----------
+    prereport : str, markdown
+
+    Returns
+    -------
+    prereport : str, html
+    """
+    out = subprocess.run(args=['pandoc'], input=prereport.encode(),
```

Comment:
like fetch/nwp.py might be good to check if pandoc is available first for helpful error

Reply:
not sure if the revision is what you had in mind here